src/models/marian.cpp

```diff
@@ -0,0 +1,205 @@
+    outputs_.push_back(encoder_outputs_.get());
+
+    State::Run(*model_.session_encoder_);
+
+    // Clear inputs and outputs for the decoder
+    ClearIO();
+
+    // Initialize the decoder inputs and outputs
+    decoder_input_ids_.name_ = model_.config_->model.decoder.inputs.input_ids.c_str();
+    decoder_input_ids_.AddMarianInputs();
+
+    next_tokens.CpuSpan()[next_tokens.size() - 1] = model_.config_->model.bos_token_id;
+
+    decoder_input_ids_.Update(next_tokens.subspan(next_tokens.size() - 1, 1));
+
+    const std::array<int64_t, 1> past_key_values_length_shape{1};
+    past_key_values_length_ = OrtValue::CreateTensor(model_.allocator_cpu_, past_key_values_length_shape, model_.session_info_.GetInputDataType(model_.config_->model.decoder.inputs.past_key_values_length));
+    input_names_.push_back(model_.config_->model.decoder.inputs.past_key_values_length.c_str());
+    inputs_.push_back(past_key_values_length_.get());
+
+    *past_key_values_length_->GetTensorMutableData<int64_t>() = -1;
+
+    attention_mask_.attention_mask_name_ = model_.config_->model.decoder.inputs.encoder_attention_mask;
+    attention_mask_.Add();
+    attention_mask_.Update(next_tokens, current_length, static_cast<int>(new_length));
+
+    auto hidden_states_type = model_.session_info_.GetInputDataType(model_.config_->model.decoder.inputs.encoder_hidden_states.c_str());
+    const std::array<int64_t, 3> encoder_hidden_states_shape{decoder_input_ids_.GetMarianInputsShape()[0], encoder_input_ids_.GetShape()[1], encoder_hidden_size};
+    encoder_hidden_states_ = OrtValue::CreateTensor(model_.p_device_inputs_->GetAllocator(), encoder_hidden_states_shape, hidden_states_type);
+
+    input_names_.push_back(model_.config_->model.decoder.inputs.encoder_hidden_states.c_str());
+    inputs_.push_back(encoder_outputs_.get());
+
+    auto rnn_states_prev_type = model_.session_info_.GetInputDataType(model_.config_->model.decoder.inputs.rnn_prev_states.c_str());
+    const std::array<int64_t, 3> rnn_states_prev_shape{3, decoder_input_ids_.GetMarianInputsShape()[0], encoder_hidden_size};
+    rnn_states_prev_ = OrtValue::CreateTensor(model_.p_device_inputs_->GetAllocator(), rnn_states_prev_shape, rnn_states_prev_type);
+
+    input_names_.push_back(model_.config_->model.decoder.inputs.rnn_prev_states.c_str());
+    auto* device_span = rnn_states_prev_->GetTensorMutableData<int32_t>();
+    size_t numel = 1;
+    for (auto dim : rnn_states_prev_shape) numel *= dim;
+    std::memset(device_span, 0, numel * sizeof(int32_t));
+    inputs_.push_back(rnn_states_prev_.get());
+
+    auto rnn_states_type = model_.session_info_.GetOutputDataType(model_.config_->model.decoder.outputs.rnn_states.c_str());
+    const std::array<int64_t, 3> rnn_states_shape{3, decoder_input_ids_.GetMarianInputsShape()[0], encoder_hidden_size};
+    rnn_states_ = OrtValue::CreateTensor(model_.p_device_inputs_->GetAllocator(), rnn_states_shape, rnn_states_type);
```

Comment:
Use Tensor instead of ortvalue. Tensor offers a way to get the device span that you can easily manipulate.